src/typ/environment.rs

```diff
@@ -466,22 +484,112 @@ impl<'a, 'b> Environment<'a, 'b> {
         }
     }
 
-    pub fn convert_unused_to_warnings(&mut self) {
-        for (name, location) in self.unused_private_types.drain() {
-            self.warnings.push(Warning::UnusedType { name, location })
+    /// Inserts an entity at the current scope for usage tracking.
+    pub fn init_usage(&mut self, name: String, kind: EntityKind, location: SrcSpan) {
+        match self
+            .entity_usages
+            .last_mut()
+            .gleam_expect("Attempted to access non-existant entity usages scope")
+            .insert(name.to_string(), (kind, location, 0))
+        {
+            // PrivateTypes can be overwritten by a constructor with the same name
+            Some((EntityKind::PrivateType, _, _)) => {}
+            Some((kind, location, 0)) => {
+                // an entity was overwritten in the top most scope without being used
+                let mut unused = HashMap::with_capacity(1);
+                let _ = unused.insert(name, (kind, location, 0));
+                self.handle_unused(unused);
+            }
+            _ => {}
         }
-        for (name, location) in self.unused_private_mixed_constructors.drain() {
-            self.warnings
-                .push(Warning::UnusedConstructor { name, location })
+    }
+
+    /// Increments an entity's usage in the current or nearest enclosing scope
+    pub fn increment_usage(&mut self, name: &str) {
+        let mut cascade = None;
+        for scope in self.entity_usages.iter_mut().rev() {
+            match scope.get_mut(name) {
+                Some((EntityKind::PrivateTypeConstructor(type_name), _, usages)) => {
+                    // If a type constructor is used, we consider
+                    // its type also used
+                    if type_name != name {
+                        cascade = Some(type_name.clone());
+                    }
+                    *usages += 1;
+                    break;
+                }
+                Some(val) => {
+                    val.2 = val.2 + 1;
+                    break;
+                }
+                None => {}
+            }
         }
+        if let Some(cascade) = cascade {
+            self.increment_usage(&cascade)
+        };
     }
 
-    pub fn type_used(&mut self, name: &str) {
-        let _ = self.unused_private_mixed_constructors.remove(name);
-        let _ = self.unused_private_types.remove(name);
+    /// Converts entities with a usage count of 0 to warnings
+    pub fn convert_unused_to_warnings(&mut self) {
+        let unused = self.entity_usages.pop().unwrap();
+        self.handle_unused(unused);
     }
 
-    pub fn value_used(&mut self, name: &str) {
-        let _ = self.unused_private_mixed_constructors.remove(name);
+    fn handle_unused(&mut self, unused: HashMap<String, (EntityKind, SrcSpan, usize)>) {
+        for val in unused {
+            match val {
+                (name, (EntityKind::ImportedType, location, 0)) => {
```

Comment:
Why is it that we track the number of times use instead of deleting the entry from the hashset if it is used? I can only see the `0` case handled, so do we care about the others?

Reply:
Because if we removed the entity from the scope on usage, the next usage would result in a variable in a wider scope getting counted as used. Essentially this has to do with value name shadowing, for example:
```
const a = 1

fn b(a) {
// | argument gets counted as used here and removed
// |   | const gets counted as used here and removed
   a + a
}
```

Reply:
That being said, a count is probably not necessary, a boolean is probably good enough.

Reply:
Ah yes, that makes sense. Thank you